api/src/main/java/io/cloudevents/CloudEvent.java

```diff
@@ -15,31 +15,62 @@
  */
 package io.cloudevents;
 
+import com.fasterxml.jackson.databind.JsonNode;
+
 import java.util.Map;
 import java.util.Optional;
 
 /**
  * An abstract event envelope
- * @param <A> The attributes type
- * @param <T> The 'data' type
  * @author fabiojose
+ * @author slinkydeveloper
  */
-public interface CloudEvent<A extends Attributes, T> {
+public interface CloudEvent {
 
     /**
      * The event context attributes
      */
-    A getAttributes();
+    Attributes getAttributes();
 
     /**
      * The event data
      */
-    Optional<T> getData();
+    Optional<String> getDataAsString() throws DataConversionException;
 
-    byte[] getDataBase64();
+    /**
+     * The event data
+     */
+    Optional<byte[]> getDataAsBytes() throws DataConversionException;
+
+    /**
+     * The event data
+     */
+    Optional<JsonNode> getDataAsJson() throws DataConversionException;
 
     /**
      * The event extensions
+     *
+     * Extensions values could be String/Number/Boolean
      */
     Map<String, Object> getExtensions();
+
+    /**
+     * Write an extension into this cloud event
+     * @param e
+     */
+    default void writeExtension(Extension e) {
```

Comment:
this brings mutability to this important interface.

It would be better to have it as `CloudEvent withExtension(Extension e)` that returns a new one

Reply:
Now that i see it, i prefer to have it removed since i designed the `CloudEvent` to be immutable

Reply:
`event.toBuilder().extension(e).build()` is an option too

Reply:
Yeah that's what i did :)